### PR TITLE
[Hackathon 7th No.55] Add `fft_conv1d` to `PaddleSpeech` -part

### PR DESCRIPTION
**Pull request by DrRyanHuang**
### PR types

New features

### PR changes

APIs

### Describe

`fft_conv1d` is much faster than `conv1d` when kernel_size >= 128

#### Implementation:

- Standard conv1D applies convolutional kernels to the input signal using a sliding window approach.
- fft_conv1d leverages the Fast Fourier Transform (FFT) to accelerate the convolution operation, ideally reducing the computational complexity from $$O(N \cdot K)$$ to $$O(N \log N)$$, where $$N$$ denotes the signal length and $$K$$ represents the kernel length.

#### Performance:

- Standard conv1D: Typically performs well with small kernels or short input signals.
Simple to implement and easy to debug.
- fft_conv1d: Generally offers performance improvements when the kernel size is large (e.g., above 256) and the stride is 1. However, the introduction of FFT and IFFT incurs additional memory overhead.

#### Result:
ce70a8c4431c0b822ed723f8f48323e3

- origin PR: https://github.com/PaddlePaddle/PaddleSpeech/pull/3900
- RFC: https://github.com/PaddlePaddle/community/pull/1017

### CODE REVIEW

File: paddlespeech/t2s/modules/fftconv1d.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+import paddle.nn as nn
+import paddle.nn.functional as F
+
+__all__ = [
+    "fft_conv1d",
+    "FFTConv1D",
+]
+
+
+def __unfold(_input, kernel_size: int, stride: int):
```
Comment (zxcd) [inline paddlespeech/t2s/modules/fftconv1d.py]: What does this api do?

File: paddlespeech/t2s/modules/fftconv1d.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    padded = F.pad(_input, (0, tgt_length - length), data_format="NCL")
+    strides: typing.List[int] = []
+    for dim in range(padded.dim()):
+        strides.append(padded.strides[dim])
+    assert strides.pop(-1) == 1, "data should be contiguous"
+    strides = strides + [stride, 1]
+    return padded.as_strided(shape + [n_frames, kernel_size], strides)
+
+
+def fft_conv1d(
```
Comment (zxcd) [inline paddlespeech/t2s/modules/fftconv1d.py:62]: use `x` replace `_input` to unified code style.

File: paddlespeech/t2s/modules/fftconv1d.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        else:
+            self.bias = None
+
+    def forward(self, _input: paddle.Tensor):
+        return fft_conv1d(_input, self.weight, self.bias, self.stride,
+                          self.padding)
+
+
+# Currently, the API unfold in Paddle is extremely slow, so __unfold is implemented 
+# using the `.strides` and `.as_strided` APIs. However, these are only supported in
```
Comment (zxcd) [inline paddlespeech/t2s/modules/fftconv1d.py:210]: paddle do not have .strides release api. Also for as_strided support >2.6, but can be combined with `view` and `reshape` .

File: paddlespeech/t2s/modules/fftconv1d.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+    def forward(self, _input: paddle.Tensor):
+        return fft_conv1d(_input, self.weight, self.bias, self.stride,
+                          self.padding)
+
+
+# Currently, the API unfold in Paddle is extremely slow, so __unfold is implemented 
+# using the `.strides` and `.as_strided` APIs. However, these are only supported in 
+# Paddle version 2.6 and above, so F.conv1d and Conv1D are used as replacements.
+version = paddle.__version__
```
Comment (zxcd) [inline paddlespeech/t2s/modules/fftconv1d.py]: version check can use `paddlespeech.utils.satisfy_paddle_version`

File: paddlespeech/t2s/modules/fftconv1d.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+if version < '2.6':
```
Comment (zxcd) [inline paddlespeech/t2s/modules/fftconv1d.py]: v2.6 and above cannot be used?

File: tests/unit/tts/test_fftconv1d.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            self.kernel_size,
+            stride=self.stride,
+            padding=self.padding,
+            bias_attr=False)
+        fft_conv1d.weight.set_value(conv1d.weight.numpy())
+        x = paddle.randn([self.batch_size, self.in_channels, self.input_length])
+        out_conv1d = conv1d(x)
+        out_fft_conv1d = fft_conv1d(x)
+        self.assertTrue(
+            np.allclose(out_conv1d.numpy(), out_fft_conv1d.numpy(), atol=1e-5))
```
Comment (zxcd) [inline tests/unit/tts/test_fftconv1d.py]: all test need 1e-6

File: paddlespeech/t2s/modules/fftconv1d.py (hunk 2 above)
Comment (DrRyanHuang) [reply paddlespeech/t2s/modules/fftconv1d.py:62]: Done

File: paddlespeech/t2s/modules/fftconv1d.py (hunk 4 above)
Comment (DrRyanHuang) [reply paddlespeech/t2s/modules/fftconv1d.py]: Done

File: tests/unit/tts/test_fftconv1d.py (hunk 6 above)
Comment (DrRyanHuang) [reply tests/unit/tts/test_fftconv1d.py]: Done

**Review by zxcd — APPROVED** [review]
LGTM

**Pull request merged**